File: Data Structures & Algorithms/house-robber/submission-1.py
class Solution:
    def rob(self, nums: List[int]) -> int:
        n = len(nums)
        if n <= 2:
            return max(nums)
        dp = [nums[0], max(nums[:2])]
        

        for i in range(2, n):
            dp.append(max(dp[i-2] + nums[i], dp[i-1]))

        return max(dp[n-1], dp[n-2])
        

File: Data Structures & Algorithms/top-k-elements-in-list/submission-1.cpp
class Solution {
public:
    vector<int> topKFrequent(vector<int>& nums, int k) {
        // 1) count frequencies
        unordered_map<int,int> freq;
        for (int x : nums) {
            freq[x]++;
        }

        // 2) min-heap (size k) of (frequency, value)
        //    keep the top-k most frequent items
        priority_queue<
            pair<int,int>,
            vector<pair<int,int>>,
            greater<pair<int,int>>
        > pq;

        for (auto &p : freq) {
            int val = p.first;
            int f   = p.second;
            pq.push({f, val});
            if (pq.size() > k) {
                pq.pop();
            }
        }

        // 3) extract results
        vector<int> res;
        // res.reserve(k);
        while (!pq.empty()) {
            res.push_back(pq.top().second);
            pq.pop();
        }
        return res;
    }
};